core/src/forwarding_stage.rs

```diff
@@ -0,0 +1,618 @@
+            }
+        }
+    }
+
+    fn forward_buffered_packets(&mut self) {
+        self.metrics.did_something |= !self.packet_container.priority_queue.is_empty();
+        self.refresh_data_budget();
+
+        // Get forwarding addresses otherwise return now.
+        let ForwardingAddresses {
+            tpu: Some(tpu),
+            tpu_vote: Some(tpu_vote),
+        } = self
+            .forward_address_getter
+            .get_forwarding_addresses(self.connection_cache.protocol())
+        else {
+            return;
+        };
+
+        let mut vote_batch = Vec::with_capacity(FORWARD_BATCH_SIZE);
+        let mut non_vote_batch = Vec::with_capacity(FORWARD_BATCH_SIZE);
+
+        // Loop through packets creating batches of packets to forward.
+        while let Some(priority_index) = self.packet_container.priority_queue.pop_max() {
+            let packet = self
+                .packet_container
+                .packets
+                .get(priority_index.index)
+                .expect("packet exists");
+
+            // If it exceeds our data-budget, drop.
+            if !self.data_budget.take(packet.meta().size) {
+                self.metrics.votes_dropped_on_data_budget +=
+                    usize::from(packet.meta().is_simple_vote_tx());
+                self.metrics.non_votes_dropped_on_data_budget +=
+                    usize::from(!packet.meta().is_simple_vote_tx());
+                self.packet_container.packets.remove(priority_index.index);
+                continue;
+            }
+
+            let packet_data_vec = packet.data(..).expect("packet has data").to_vec();
+
+            if packet.meta().is_simple_vote_tx() {
+                vote_batch.push(packet_data_vec);
+                if vote_batch.len() == vote_batch.capacity() {
+                    self.send_vote_batch(tpu_vote, &mut vote_batch);
+                }
+            } else {
+                non_vote_batch.push((packet_data_vec, tpu));
+                if non_vote_batch.len() == non_vote_batch.capacity() {
+                    self.send_non_vote_batch(&mut non_vote_batch);
+                }
+            }
+        }
+
+        // Send out remaining packets
+        if !vote_batch.is_empty() {
+            self.metrics.votes_forwarded += vote_batch.len();
+            self.send_vote_batch(tpu_vote, &mut vote_batch);
+        }
+        if !non_vote_batch.is_empty() {
+            self.metrics.non_votes_forwarded += non_vote_batch.len();
+            self.send_non_vote_batch(&mut non_vote_batch);
+        }
+    }
+
+    /// Re-fill the data budget if enough time has passed
+    fn refresh_data_budget(&self) {
+        const INTERVAL_MS: u64 = 100;
+        // 12 MB outbound limit per second
+        const MAX_BYTES_PER_SECOND: usize = 12_000_000;
+        const MAX_BYTES_PER_INTERVAL: usize = MAX_BYTES_PER_SECOND * INTERVAL_MS as usize / 1000;
+        const MAX_BYTES_BUDGET: usize = MAX_BYTES_PER_INTERVAL * 5;
```

Comment:
Please use Duration for timings.

Reply:
`DataBudget` is re-used in multiple places. Fine to do it after this PR, but the interface for that `update` fn takes in an integer.